TPC Cluster Removal and Compression Branch

davidrohr opened the pull request:
Not really WIP in the sense that is already fully working so far, but track model, sorting and building differences, as well as stripping bits from width / charge is not yet present.

Will add this step by step and then merge as a whole.

No connection to entry compression yet, just creates a container with entropy-reduced arrays of cluster properties, which should be entropy encoded afterwards. Could be merged eventually to avoid the memory overhead of keeping the clusters twice.

shahor02 commented on DataFormats/Detectors/TPC/include/DataFormatsTPC/CompressedClusters.h:
```diff
@@ -0,0 +1,61 @@
+  unsigned short* nTrackClusters = nullptr;  // [nTracks]
+  unsigned int* nSliceRowClusters = nullptr; // [36 * 152]
+
+  ClassDefNV(CompressedClusters, 1);
```
This will generate a wrong streamer since the persistent arrays dimensions are not given according to rootcling rules e.g. ``char* qPtA = nullptr; //[nTracks]``. If this class never fill be root-serialized, better use ``ClassDefNV(CompressedClusters, 0);``, otherwise remove space in ``// [`` and use only its data members of constants (no expressions) for dimensions.

davidrohr replied:
@shahor02 : Thanks, I was actually not sure if this would work and wanted to check it at some time. I'll adapt it.
Do you by chance know if it works with constexpr static data members as constants? I didn't find anything about it in the ROOT documentation.

shahor02 replied:
@davidrohr : no, one should either use a persistent data member to indicate the expected size:
````
unsigned int mNSlots = 36 * 152;
unsigned int *nSliceRowClusters = nullptr; //[mNSlots]
````
or fixed size array e.g.
````
static constexp int  NSlots = 36 * 152;
unsigned int nSliceRowClusters[NSlots];
````

davidrohr commented:
ROOT streamers fixed, tested and working.
GPU processing tested and working.
travis issues fixed

davidrohr commented:
Truncation for charge / width is in now, as well as storing of differences.
Still need to add different sort modi for differences and finally track model compression.

If there are no more comments, I'll merge this after I fixed the merge conflict ( just to prevent further conflicts). It is fully working, only some additional features missing that can be added later.

shahor02 commented:
No comments from my side (except that I would define values for tpcCompressionModes in enum instead of hardcoding).

davidrohr commented:
> No comments from my side (except that I would define values for tpcCompressionModes in enum instead of hardcoding).

Sure, will do that eventually. Right now I am not yet sure what kind of settings I need.

davidrohr commented:
Everything is in now except the track model. I rebased after the ITS namespace change and fixed the collissions and also the ITS GPU tracking compilation, which was broken.
Will merge once the C.I. is green.
